### PR TITLE
Cancel research variants request

### LINKED ISSUE #3107: Cancel request to load research variants from case page
It's useful if one clicks and requests research variants by mistake. Otherwise an admin must connect to the database and edit the case document in there.

### PR DESCRIPTION
**Pull request by northwestwitch**
This PR:
- Adds a button to allow a user to cancel a request for research variants for a case  (fix #3107 )
- Fixes the style of rerun and research requests from this:

image

to this:

image

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. `ssh your.email@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
</details>

**How to test**:
1. Locally, on demo case, request research variants
2. Cancel the request clicking on

image

**Expected outcome**:
- [ ] The research status should be reset
- [ ] An event should be displayed on case page, sayin that the user has canceled a research variants request

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

**Comment by northwestwitch** [comment]
Tested locally: 🆗 

image

image

**Comment by northwestwitch** [comment]
This PR is fixing that issue of a request for research variants from a case that has been solved. I don't find the issue any more, where was it @Mropat @moahaegglund ?

**Comment by moahaegglund** [comment]
> This PR is fixing that issue of a request for research variants from a case that has been solved. I don't find the issue any more, where was it @Mropat @moahaegglund ?

Here: [https://github.com/Clinical-Genomics/IT-issues/issues/458](https://github.com/Clinical-Genomics/IT-issues/issues/458)

**Comment by northwestwitch** [comment]
> Here: [Clinical-Genomics/IT-issues#458](https://github.com/Clinical-Genomics/IT-issues/issues/458)

Thanks!

**Comment by northwestwitch** [comment]
I would like to include also this one in the new release. Could you review when have time @dnil @mikaell ?

**Comment by dnil** [comment]
Hihi, I had just switched to it! 😸

**Comment by northwestwitch** [comment]
> Hihi, I had just switched to it! 😸

You are on a reviewing spree

File: tests/server/blueprints/cases/test_cases_views.py [hunk 1]
```diff
@@ -91,6 +91,68 @@ def test_reanalysis(app, institute_obj, case_obj, mocker, mock_redirect):
         assert resp.status_code == 302
 
 
+def test_research(app, institute_obj, case_obj, mocker, mock_redirect):
```
Comment (dnil) [inline tests/server/blueprints/cases/test_cases_views.py:94]: 🥇

**Comment by dnil** [comment]
🆗 event
Screenshot 2022-02-04 at 11 21 40
🆗 CLI exports correctly before and after cancel
Screenshot 2022-02-04 at 11 22 56

File: scout/constants/case_tags.py [hunk 2]
```diff
@@ -71,6 +71,7 @@
     "cancel_sanger": "cancelled sanger order for",
     "archive": "archived",
     "open_research": "opened research mode for",
+    "close_research": "canceled research mode for",
```
Comment (dnil) [inline scout/constants/case_tags.py:74]:
```suggestion
    "reset_research": "canceled research mode request for",
```

Comment (dnil) [reply scout/constants/case_tags.py:74]: It's not really requesting to close research mode. Which could totally be a thing, but for another issue! Reset is good.

File: scout/adapter/mongo/case_events.py [hunk 3]
```diff
@@ -282,6 +282,41 @@ def open_research(self, institute, case, user, link):
         LOG.debug("Case updated")
         return updated_case
 
+    def close_research(self, institute, case, user, link):
```
Comment (dnil) [inline scout/adapter/mongo/case_events.py:285]:
```suggestion
    def reset_research(self, institute, case, user, link):
```

File: scout/adapter/mongo/case_events.py [hunk 4]
```diff
@@ -282,6 +282,41 @@ def open_research(self, institute, case, user, link):
...
+
+        LOG.info("Creating event for closing research for case" " {0}".format(case["display_name"]))
+
+        self.create_event(
+            institute=institute,
+            case=case,
+            user=user,
+            link=link,
+            category="case",
+            verb="close_research",
```
Comment (dnil) [inline scout/adapter/mongo/case_events.py:306]:
```suggestion
            verb="reset_research",
```

File: scout/server/blueprints/cases/views.py [hunk 5]
```diff
@@ -842,6 +842,15 @@ def research(institute_id, case_name):
     return redirect(request.referrer)
 
 
+@cases_bp.route("/<institute_id>/<case_name>/reset_research", methods=["GET"])
+def reset_research(institute_id, case_name):
+    institute_obj, case_obj = institute_and_case(store, institute_id, case_name)
+    user_obj = store.user(current_user.email)
+    link = url_for(".case", institute_id=institute_id, case_name=case_name)
+    store.close_research(institute_obj, case_obj, user_obj, link)
```
Comment (dnil) [inline scout/server/blueprints/cases/views.py:850]:
```suggestion
    store.reset_research(institute_obj, case_obj, user_obj, link)
```

File: tests/server/blueprints/cases/test_cases_views.py [hunk 6]
```diff
@@ -91,6 +91,68 @@ def test_reanalysis(app, institute_obj, case_obj, mocker, mock_redirect):
...
+            url_for(
+                "cases.reset_research",
+                institute_id=institute_obj["internal_id"],
+                case_name=case_obj["display_name"],
+            ),
+        )
+        # THEN the updated case should have research_requested set to False
+        assert store.case_collection.find_one({"research_requested": False})
+        # AND a relative event should have been created in the evens collection
+        assert store.event_collection.find_one({"verb": "close_research"})
```
Comment (dnil) [inline tests/server/blueprints/cases/test_cases_views.py:153]:
```suggestion
        assert store.event_collection.find_one({"verb": "reset_research"})
```

**Review by dnil — APPROVED** [review]
Useful feature!

You have a few (false positive) unused from Vulture - perhaps #noqa those lines when you touch it!

There is more than one way to do it. We have endpoints with POST and an argument to cancel / activate and just toggle endpoints. This now has a POST to activate and a GET to cancel. It's all fine, though something that would likely be nice to refactor with a more generic event handler in the future.

I think reset_research was a good name; close_research makes me think someone is about to remove an already existing set of research variants.

**Pull request merged**